src/bin/commands/demux.rs

```diff
@@ -88,17 +119,17 @@ impl ReadSet {
             writer,
             read_num,
             self.header.as_slice(),
-            self.sample_barcode_segments.as_slice(),
-            self.molecular_barcode_segments.as_slice(),
+            self.sample_barcode_segments(),
+            self.molecular_barcode_segments(),
         )
     }
 
-    fn write_header_internal<W: Write>(
+    fn write_header_internal<'a, W: Write>(
         writer: &mut W,
         read_num: usize,
         header: &[u8],
-        sample_barcode_segments: &[FastqSegment],
-        molecular_barcode_segments: &[FastqSegment],
+        sample_barcode_segments: impl Iterator<Item = &'a FastqSegment>,
+        mut molecular_barcode_segments: impl Iterator<Item = &'a FastqSegment>,
```

Comment:
because the `next()` function (on molecular_barcode_segments, and not on sample_barcode) we use below requires a mutable reference and the compiler won't let me not make it mutable.